contracts/JoepegAuctionHouse.sol

```diff
@@ -0,0 +1,960 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IERC20, SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC721} from "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {ReentrancyGuardUpgradeable} from "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+
+import {ICurrencyManager} from "./interfaces/ICurrencyManager.sol";
+import {IProtocolFeeManager} from "./interfaces/IProtocolFeeManager.sol";
+import {IRoyaltyFeeManager} from "./interfaces/IRoyaltyFeeManager.sol";
+import {IWAVAX} from "./interfaces/IWAVAX.sol";
+
+error JoepegAuctionHouse__AuctionAlreadyExists();
+error JoepegAuctionHouse__CurrencyMismatch();
+error JoepegAuctionHouse__ExpectedNonnullAddress();
+error JoepegAuctionHouse__InvalidDuration();
+error JoepegAuctionHouse__NoAuctionExists();
+error JoepegAuctionHouse__OnlyAuctionCreatorCanCancel();
+error JoepegAuctionHouse__TransferAVAXFailed();
+error JoepegAuctionHouse__UnsupportedCurrency();
+
+error JoepegAuctionHouse__EnglishAuctionCannotBidOnEndedAuction();
+error JoepegAuctionHouse__EnglishAuctionCannotCancelWithExistingBid();
+error JoepegAuctionHouse__EnglishAuctionCannotSettleWithoutBid();
+error JoepegAuctionHouse__EnglishAuctionCreatorCannotPlaceBid();
+error JoepegAuctionHouse__EnglishAuctionInsufficientBidAmount();
+error JoepegAuctionHouse__EnglishAuctionInvalidMinBidIncrementPct();
+error JoepegAuctionHouse__EnglishAuctionInvalidRefreshTime();
+error JoepegAuctionHouse__EnglishAuctionOnlyCreatorCanSettleBeforeEndTime();
+
+error JoepegAuctionHouse__DutchAuctionCreatorCannotSettle();
+error JoepegAuctionHouse__DutchAuctionInsufficientAmountToSettle();
+error JoepegAuctionHouse__DutchAuctionInvalidStartEndPrice();
+
+/**
+ * @title JoepegAuctionHouse
+ * @notice An auction house that supports running English and Dutch auctions on ERC721 tokens
+ */
+contract JoepegAuctionHouse is
+    Initializable,
+    OwnableUpgradeable,
+    ReentrancyGuardUpgradeable
+{
+    using SafeERC20 for IERC20;
+
+    struct DutchAuction {
+        address creator;
+        address currency;
+        uint256 startPrice;
+        uint256 endPrice;
+        uint256 startTime;
+        uint256 endTime;
+        uint256 dropInterval;
+    }
+
+    struct EnglishAuction {
+        address creator;
+        address currency;
+        address lastBidder;
+        uint256 lastBidPrice;
+        uint256 endTime;
+        uint256 startPrice;
+    }
+
+    uint256 public immutable PERCENTAGE_PRECISION = 10000;
+
+    address public WAVAX;
```

Comment:
this sounds good to me but had quick question — I thought upgradeable contracts couldn't have constructors. is this safe to do?

Reply:
Upgradeable contracts can have constructor, it's just that the values that aren't stored within the bytecode (any values that aren't immutable or constants), will not have the expected value because it will load the proxy's storage slot instead of the implementation ones.
That also means that it can't be changed. For example, if you use clone pattern, all cloned contract would have the same values for the immutable and constant because the constructor would not be run again.